pkg/engine/internal/executor/cast.go

```diff
@@ -0,0 +1,225 @@
+package executor
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/apache/arrow-go/v18/arrow"
+	"github.com/apache/arrow-go/v18/arrow/array"
+	"github.com/apache/arrow-go/v18/arrow/memory"
+	"github.com/dustin/go-humanize"
+
+	"github.com/grafana/loki/v3/pkg/engine/internal/semconv"
+	"github.com/grafana/loki/v3/pkg/engine/internal/types"
+)
+
+func castFn(operation types.UnaryOp) UnaryFunction {
+	return UnaryFunc(func(input ColumnVector, allocator memory.Allocator) (ColumnVector, error) {
+		arr := input.ToArray()
+		defer arr.Release()
+
+		sourceCol, ok := arr.(*array.String)
+		if !ok {
+			return nil, fmt.Errorf("expected column to be of type string, got %T", arr)
+		}
+
+		// Get conversion function and process values
+		conversionFn := getConversionFunction(operation)
+		castCol, errTracker := castValues(sourceCol, conversionFn, allocator)
+		defer castCol.Release()
+
+		// Build error columns if needed
+		errorCol, errorDetailsCol := errTracker.buildArrays()
+		defer errTracker.releaseBuilders()
+		if errTracker.hasErrors {
+			defer errorCol.Release()
+			defer errorDetailsCol.Release()
+		}
+
+		// Build output schema and record
+		fields := buildOutputFields(errTracker.hasErrors)
+		result, err := buildResult(castCol, errorCol, errorDetailsCol, fields)
+		if err != nil {
+			return nil, err
+		}
+		return &ArrayStruct{
+			array: result,
+			ct:    types.ColumnTypeGenerated,
+			rows:  input.Len(),
+		}, nil
+	})
+}
+
+type conversionFn func(value string) (float64, error)
+
+func getConversionFunction(operation types.UnaryOp) conversionFn {
+	switch operation {
+	case types.UnaryOpCastBytes:
+		return convertBytes
+	case types.UnaryOpCastDuration:
+		return convertDuration
+	default:
+		return convertFloat
+	}
+}
+
+func castValues(
+	sourceCol *array.String,
+	conversionFn conversionFn,
+	allocator memory.Allocator,
+) (arrow.Array, *errorTracker) {
+	castBuilder := array.NewFloat64Builder(allocator)
+	defer castBuilder.Release()
+
+	tracker := newErrorTracker(allocator)
+
+	for i := 0; i < sourceCol.Len(); i++ {
+		if sourceCol.IsNull(i) {
+			castBuilder.AppendNull()
+			tracker.recordSuccess()
+		} else {
+			valueStr := sourceCol.Value(i)
+			if val, err := conversionFn(valueStr); err == nil {
+				castBuilder.Append(val)
+				tracker.recordSuccess()
+			} else {
+				// Use 0.0 as default for errors, for backwards compatibility with old engine
+				castBuilder.Append(0.0)
+				tracker.recordError(i, err)
+			}
+		}
+	}
+
+	return castBuilder.NewArray(), tracker
+}
+
+func buildOutputFields(
+	hasErrors bool,
+) []arrow.Field {
+	fields := make([]arrow.Field, 0, 3)
+
+	// Add value field
+	fields = append(fields, semconv.FieldFromIdent(semconv.ColumnIdentValue, false))
```

Comment:
While we never have a NULL value in this column (rather `0`), we still make any columns nullable.

Even the builting message and timestamp columns, which are truly not nullable, are defined as nullable columns :/

Reply:
On the other hand, we nowhere enforce nullable=false